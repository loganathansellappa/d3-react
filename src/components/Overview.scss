.overview {
  display: flex;
  flex-direction: column;
  h2 {
    font-size: 24px;
    margin-bottom: 20px;
  }
}