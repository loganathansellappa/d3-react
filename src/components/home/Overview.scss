.overview {
  display: flex;
  flex-direction: column;

  h2 {
    font-size: 24px;
    margin-bottom: 20px;
  }
  .overview__content {
    display: flex;
    flex-direction: row;
    table {
      flex: 0 0 60%;
      margin-right: 50px;
    }
    .chart {
      margin-top: 100px;
      flex: 0 0 40%;
    }
  }
}
