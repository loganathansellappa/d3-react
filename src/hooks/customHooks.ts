import axios from "axios";
import { useQuery } from "react-query";
import {
  NormalizedCompanySchema,
  BalanceSheetSchema,
  IncomeStatementSchema,
  ApiLimitSchema,
  TimeSeriesDailySchema,
} from "../@types/zodSchema";
import { z } from "zod";
import {serverData, transformTimeSeriesDaily} from "../utils/HelperUtils";

const TransformedTimeSeriesDailySchema = TimeSeriesDailySchema.transform(
  (data) => transformTimeSeriesDaily(data["Time Series (Daily)"]),
);
const fetchData = async (fn: string = "OVERVIEW") => {
  const url = `${serverData().baseUrl}?function=${fn}&symbol=IBM&apikey=${
    serverData().apiKey
  }`;
  const response = await axios.get(url);
  const schemaMap: Record<string, z.Schema<unknown>> = {
    OVERVIEW: NormalizedCompanySchema,
    BALANCE_SHEET: BalanceSheetSchema,
    INCOME_STATEMENT: IncomeStatementSchema,
    TIME_SERIES_DAILY: TransformedTimeSeriesDailySchema,
    LIMIT_EXCEEDED: ApiLimitSchema,
    // Add more functions and schemas as needed
  };
  if (response.data.Information) {
    throw new Error(response.data.Information);
  }
  return schemaMap[fn].parse(response.data);
};

export function useApi(fn: string = "OVERVIEW") {
  return useQuery({
    queryKey: fn,
    queryFn: () => fetchData(fn),
  });
}


