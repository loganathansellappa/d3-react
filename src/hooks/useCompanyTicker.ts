import { useLocation } from "react-router-dom";

export const useCompanyTicker = (): string => {
  const location = useLocation();
  const searchParams = new URLSearchParams(location.search);
  const sym = searchParams.get("sym");
  return sym || "IBM"; // Provide the default value here
};
