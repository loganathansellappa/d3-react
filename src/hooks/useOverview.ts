import axios from 'axios';
import { useQuery } from 'react-query';
import {
    NormalizedCompanySchema,
    BalanceSheetSchema,
    IncomeStatementSchema,
    ApiLimitSchema,
    TimeSeriesDailySchema, TimeSeriesDataType
} from "../@types/zodSchema";
import {z} from "zod";

const TransformedTimeSeriesDailySchema = TimeSeriesDailySchema.transform(
    (data) => transformTimeSeriesDaily(data["Time Series (Daily)"]),
    (data) => data
);
const fetchData = async (fn: string = 'OVERVIEW') => {

    const response = await axios.get(
        `https://www.alphavantage.co/query?function=${fn}&symbol=IBM&apikey=demo`
    );
    const schemaMap: Record<string, z.Schema<unknown>> = {
        OVERVIEW: NormalizedCompanySchema,
        BALANCE_SHEET: BalanceSheetSchema,
        INCOME_STATEMENT: IncomeStatementSchema,
        TIME_SERIES_DAILY: TransformedTimeSeriesDailySchema,
        LIMIT_EXCEEDED: ApiLimitSchema,
        // Add more functions and schemas as needed
    };
    if (response.data.Information) {
        throw new Error(response.data.Information);
    }
    return schemaMap[fn].parse(response.data);
};


/*
    Added default value, we can extend this hook with any symbol
 */

export function useApi(fn: string = 'OVERVIEW') {
    return useQuery({
        queryKey: fn,
        queryFn: () => fetchData(fn),
    });
}

const transformTimeSeriesDaily = (data: Record<string, TimeSeriesDataType>) => {

    const keyMapping: Record<string, string> = {
        '1. open': 'open',
        '2. high': 'high',
        '3. low': 'low',
        '4. close': 'close',
        '5. volume': 'volume',
    };

    const transformedData: Record<string, Record<string, string>> = {};

    for (const date in data) {

        const innerObject = data[date];
        const transformedInnerObject: Record<string, string> = {};

        for (const oldKey in innerObject) {
            if (keyMapping[oldKey]) {
                innerObject[oldKey] = +innerObject[oldKey];
                transformedInnerObject[keyMapping[oldKey]] = innerObject[oldKey];
            } else {
                transformedInnerObject[oldKey] = innerObject[oldKey];
            }
        }

        transformedData[date] = transformedInnerObject;
    }

    return transformedData;
};

