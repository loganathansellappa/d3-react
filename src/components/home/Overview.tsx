import React from "react";
import "./Overview.scss";
import { useApi } from "../../hooks/customHooks";
import LoadingSpinner from "../loader/LoadingSpinner";
import { Chart } from "../charts/Chart";

export const Overview: React.FC = () => {
  const { data, isLoading, isError, error } = useApi("OVERVIEW");

  if (isLoading) {
    return <LoadingSpinner />;
  }

  if (isError && error instanceof Error) {
    return <div>Error fetching data {error.message}</div>;
  }

  if (data) {
    return (
      <div className="overview">
        <h2>Overview</h2>
        <div className="overview__content">
          <table>
            <thead>
              <tr>
                <th>Property</th>
                <th>Value</th>
              </tr>
            </thead>
            <tbody>
              {Object.entries(data).map(([property, value]) => (
                <tr key={property}>
                  <td className={"bold-property"}>{property}</td>
                  <td>{value}</td>
                </tr>
              ))}
            </tbody>
          </table>
          <div className="chart">
            <Chart></Chart>
          </div>
        </div>
      </div>
    );
  }
  return <div>Loading...</div>;
};
