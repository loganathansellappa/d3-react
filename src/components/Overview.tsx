import React from "react";
import "./Overview.scss";
import { useApi } from "../hooks/useOverview";
import LoadingSpinner from "./LoadingSpinner";
import { Chart } from "./Chart";

export const Overview: React.FC = () => {
  const { data, isLoading, isError, error } = useApi("OVERVIEW");

  if (isLoading) {
    return <LoadingSpinner />;
  }

  if (isError && error instanceof Error) {
    return <div>Error fetching data {error.message}</div>;
  }

  if (data) {
    return (
      <div className="overview">
        <h2>Overview</h2>
        <table>
          <thead>
            <tr>
              <th>Property</th>
              <th>Value</th>
            </tr>
          </thead>
          <tbody>
            {Object.entries(data).map(([property, value]) => (
              <tr key={property}>
                <td className={"bold-property"}>{property}</td>
                <td>{value}</td>
              </tr>
            ))}
          </tbody>
        </table>
        <Chart></Chart>
      </div>
    );
  }
  return <div>Loading...</div>;
};
