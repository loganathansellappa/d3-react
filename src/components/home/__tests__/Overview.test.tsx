import { QueryClient, QueryClientProvider, UseQueryResult } from "react-query";
import { Overview } from "../Overview";
import { mockOverviewResponse, renderWithRouter } from "../../../utils/test";
import * as Hooks from "../../../hooks/customHooks";

describe("Overview", () => {
  let element: { container: HTMLElement };
  beforeEach(() => {
    jest.spyOn(Hooks, "useApi").mockReturnValue({
      data: mockOverviewResponse(),
      isError: false,
      isLoading: false,
      refetch: jest.fn(),
    } as unknown as UseQueryResult<unknown, unknown>);
    const queryClient = new QueryClient();
    element = renderWithRouter(
      <QueryClientProvider client={queryClient}>
        <Overview />
      </QueryClientProvider>,
    );
  });

  test("renders Overview component", () => {
    expect(element.container.firstChild).toMatchSnapshot();
  });

  test("displays response data and chart", () => {
    expect(element.container.textContent).toContain(
      "International Business Machines",
    );
    expect(element.container.textContent).toContain("Chart");
  });
});
