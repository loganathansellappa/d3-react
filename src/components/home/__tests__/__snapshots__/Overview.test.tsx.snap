// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overview renders Overview component 1`] = `
<div
  class="overview"
>
  <h2>
    Overview
  </h2>
  <div
    class="overview__content"
  >
    <table>
      <thead>
        <tr>
          <th>
            International Business Machines
          </th>
          <th />
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="bold-property"
          >
            Symbol
          </td>
          <td>
            IBM
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            AssetType
          </td>
          <td>
            Common Stock
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            Name
          </td>
          <td>
            International Business Machines
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            Description
          </td>
          <td>
            International Business Machines Corporation (IBM) is an American multinational technology company headquartered in Armonk, New York, with operations in over 170 countries. The company began in 1911, founded in Endicott, New York, as the Computing-Tabulating-Recording Company (CTR) and was renamed International Business Machines in 1924. IBM is incorporated in New York. IBM produces and sells computer hardware, middleware and software, and provides hosting and consulting services in areas ranging from mainframe computers to nanotechnology. IBM is also a major research organization, holding the record for most annual U.S. patents generated by a business (as of 2020) for 28 consecutive years. Inventions by IBM include the automated teller machine (ATM), the floppy disk, the hard disk drive, the magnetic stripe card, the relational database, the SQL programming language, the UPC barcode, and dynamic random-access memory (DRAM). The IBM mainframe, exemplified by the System/360, was the dominant computing platform during the 1960s and 1970s.
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            CIK
          </td>
          <td>
            51143
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            Exchange
          </td>
          <td>
            NYSE
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            Currency
          </td>
          <td>
            USD
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            Country
          </td>
          <td>
            USA
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            Sector
          </td>
          <td>
            TECHNOLOGY
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            Industry
          </td>
          <td>
            COMPUTER & OFFICE EQUIPMENT
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            Address
          </td>
          <td>
            1 NEW ORCHARD ROAD, ARMONK, NY, US
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            FiscalYearEnd
          </td>
          <td>
            December
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            LatestQuarter
          </td>
          <td>
            2023-09-30
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            MarketCapitalization
          </td>
          <td>
            133799322000
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            EBITDA
          </td>
          <td>
            13663000000
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            PERatio
          </td>
          <td>
            18.91
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            PEGRatio
          </td>
          <td>
            1.276
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            BookValue
          </td>
          <td>
            25.28
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            DividendPerShare
          </td>
          <td>
            6.62
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            DividendYield
          </td>
          <td>
            0.0457
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            EPS
          </td>
          <td>
            7.75
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            RevenuePerShareTTM
          </td>
          <td>
            67.3
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            ProfitMargin
          </td>
          <td>
            0.113
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            OperatingMarginTTM
          </td>
          <td>
            0.145
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            ReturnOnAssetsTTM
          </td>
          <td>
            0.0455
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            ReturnOnEquityTTM
          </td>
          <td>
            0.328
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            RevenueTTM
          </td>
          <td>
            61170999000
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            GrossProfitTTM
          </td>
          <td>
            32688000000
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            DilutedEPSTTM
          </td>
          <td>
            7.75
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            QuarterlyEarningsGrowthYOY
          </td>
          <td>
            0.126
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            QuarterlyRevenueGrowthYOY
          </td>
          <td>
            0.046
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            AnalystTargetPrice
          </td>
          <td>
            137.35
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            TrailingPE
          </td>
          <td>
            18.91
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            ForwardPE
          </td>
          <td>
            15.55
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            PriceToSalesRatioTTM
          </td>
          <td>
            2.108
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            PriceToBookRatio
          </td>
          <td>
            6.75
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            EVToRevenue
          </td>
          <td>
            2.969
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            EVToEBITDA
          </td>
          <td>
            25.81
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            Beta
          </td>
          <td>
            0.771
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            52WeekHigh
          </td>
          <td>
            151.93
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            52WeekLow
          </td>
          <td>
            118.71
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            50DayMovingAverage
          </td>
          <td>
            143.62
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            200DayMovingAverage
          </td>
          <td>
            135.29
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            SharesOutstanding
          </td>
          <td>
            911006000
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            DividendDate
          </td>
          <td>
            2023-09-09
          </td>
        </tr>
        <tr>
          <td
            class="bold-property"
          >
            ExDividendDate
          </td>
          <td>
            2023-11-08
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="chart"
    >
      <div
        class="chart-wrapper"
      >
        <h2
          class="title"
        >
          Daily Data
        </h2>
        <div
          class="button-container"
        >
          <button
            class="button button-active"
          >
            Area Chart
          </button>
          <button
            class="button "
          >
            Candle Chart
          </button>
        </div>
        <div
          class="chart-container"
        >
          <svg
            viewBox="0 0 1600 800"
          >
            <g
              transform="translate(10,100)"
            />
            <g
              fill="none"
              font-family="sans-serif"
              font-size="10"
              text-anchor="middle"
              transform="translate(0,690)"
            >
              <path
                class="domain"
                d="M0.5,6V0.5H1530.5V6"
                stroke="currentColor"
              />
            </g>
            <g
              fill="none"
              font-family="sans-serif"
              font-size="10"
              text-anchor="start"
              transform="translate(1530,0)"
            >
              <path
                class="domain"
                d="M6,690.5H0.5V0.5H6"
                stroke="currentColor"
              />
            </g>
            <path
              class="area"
              d="MNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690LNaN,690Z"
              style="fill: #007bff; opacity: 0.5;"
            />
            <path
              class="line"
              d="MNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaNLNaN,NaN"
              fill="none"
              stroke="#007bff"
              stroke-width="1"
            />
            <circle
              fill="red"
              opacity="0.7"
              r="0"
              style="stroke: white; pointer-events: none;"
            />
            <line
              class="crosshair-x"
              id="crosshair-x"
              stroke="gray"
              stroke-dasharray="3,3"
            />
            <line
              class="crosshair-y"
              id="crosshair-y"
              stroke="gray"
              stroke-dasharray="3,3"
            />
            <rect
              class="background-rect"
              height="690"
              width="1530"
            />
          </svg>
          <div
            class="tooltip"
          />
          <div
            class="tooltip"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
